### PR TITLE
CC-184 | Many optional parameters in processor

### PR DESCRIPTION
**Pull request by urbanusjam**

### CODE REVIEW

File: src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    protected RuleKey getRuleKey() {
+        return RULE_KEY;
+    }
+
+    private static final int DEFAULT_MAX_ALLOWED = 6;
+
+    /**
+     * The maximum number of simple-type arguments allowed in a processor.
+     */
+    @RuleProperty(key = "maxArgumentsAllowed", defaultValue = "" + DEFAULT_MAX_ALLOWED, description = "The maximum number of complex-type arguments allowed in a method annotated with @Processor")
```
Comment (marianogoldman) [inline src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java]: Please update description.
Remember that description field supports markdown, so feel free to play with those ;-)

File: src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    protected int maxArgumentsAllowed = DEFAULT_MAX_ALLOWED;
+
+    @Override
+    protected void verifyProcessor(@NonNull MethodTree tree, final @NonNull IdentifierTree processorAnnotation) {
+        final long count = Iterables.size(Iterables.filter(tree.parameters(), ClassParserUtils.getSimpleTypePredicate(imports)));
```
Comment (marianogoldman) [inline src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java]: Hum... actually, we're not really checking that @Optional is there. We need to add that check. I would add a compound predicate (Predicates.and(predicate1, predicate2)) where predicate one is the one you're using now, and the other would be that the parameter contains @Optional annotation.

File: src/test/files/java/NumberOfSimpleAndOptionalArgumentsCheck.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    @Processor
+    public void aMethod() {
+    }
+
+    @Processor
+    public void aMethod(SompleComplexType s1, String s2, String s3, String s4, String s5) {
+    }
+
+    @Processor
+    public void aMethod(SomeComplexType s1, SomeComplexType s2, SomeComplexType s3, SomeComplexType s4, SomeComplexType s5) {
```
Comment (marianogoldman) [inline src/test/files/java/NumberOfSimpleAndOptionalArgumentsCheck.java]: Let's add another test that includes @Optional so we can test this for real. Please also include the import for the annotation.

File: src/test/resources/org/mule/tools/devkit/sonar/checks/pom/ScopeProvidedInMuleDependenciesCheckTest-pom.xml [hunk 4]
```diff
@@ -18,6 +18,7 @@
             <groupId>org.mule.tools.devkit</groupId>
             <artifactId>mule-devkit-annotations</artifactId>
             <version>3.7.2</version>
+            <scope>compile</scope>
```
Comment (marianogoldman) [inline src/test/resources/org/mule/tools/devkit/sonar/checks/pom/ScopeProvidedInMuleDependenciesCheckTest-pom.xml]: Please revert this. The idea is one pom that checks that there's no explicit scope (so it defaults to compile). And there is another test for explicitly defined compile. This is the former one.

File: src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java (hunk 1 above)
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/checks/java/NumberOfSimpleAndOptionalArgumentsCheck.java]: It's the copy/paste curse.

File: src/test/resources/org/mule/tools/devkit/sonar/checks/pom/ScopeProvidedInMuleDependenciesCheckTest-pom.xml (hunk 4 above)
Comment (urbanusjam) [inline src/test/resources/org/mule/tools/devkit/sonar/checks/pom/ScopeProvidedInMuleDependenciesCheckTest-pom.xml]: Will do. I was doing some test locally and forgot to remove it.

**Pull request merged**